atlasdb-workload-server/src/main/java/com/palantir/atlasdb/workload/invariant/SnapshotInvariantVisitor.java

```diff
@@ -0,0 +1,123 @@
+        Optional<Integer> expected = fetchValueFromView(
+                        readTransactionAction.table(), readTransactionAction.cell(), readView)
+                .value();
+        if (!expected.equals(readTransactionAction.value())) {
+            return Optional.of(InvalidWitnessedTransactionAction.of(
+                    readTransactionAction, MismatchedValue.of(readTransactionAction.value(), expected)));
+        }
+        return Optional.empty();
+    }
+
+    @Override
+    public Optional<InvalidWitnessedTransactionAction> visit(WitnessedWriteTransactionAction writeTransactionAction) {
+        Optional<InvalidWitnessedTransactionAction> invalidAction = checkForWriteWriteConflicts(
+                        writeTransactionAction.table(), writeTransactionAction.cell())
+                .map(mismatchedValue -> InvalidWitnessedTransactionAction.of(writeTransactionAction, mismatchedValue));
+
+        applyWrites(
+                writeTransactionAction.table(),
+                writeTransactionAction.cell(),
+                Optional.of(writeTransactionAction.value()));
+        return invalidAction;
+    }
+
+    @Override
+    public Optional<InvalidWitnessedTransactionAction> visit(WitnessedDeleteTransactionAction deleteTransactionAction) {
+        Optional<InvalidWitnessedTransactionAction> invalidAction = checkForWriteWriteConflicts(
+                        deleteTransactionAction.table(), deleteTransactionAction.cell())
+                .map(mismatchedValue -> InvalidWitnessedTransactionAction.of(deleteTransactionAction, mismatchedValue));
+
+        applyWrites(deleteTransactionAction.table(), deleteTransactionAction.cell(), Optional.empty());
+        return invalidAction;
+    }
+
+    /**
+     * Applies writes to both our read and latest table view.
+     * Latest table view needs to be updated to apply our writes from our transaction,
+     * while the read view is updated to reflect local writes.
+     */
+    private void applyWrites(String tableName, WorkloadCell cell, Optional<Integer> value) {
+        readView.with(table ->
+                table.put(TableAndWorkloadCell.of(tableName, cell), ValueAndTimestamp.of(value, startTimestamp)));
+        latestView.with(table ->
+                table.put(TableAndWorkloadCell.of(tableName, cell), ValueAndTimestamp.of(value, startTimestamp)));
+    }
+
+    /**
+     * Checks that the value we are writing has not changed from our read view and the latest view. If it has,
+     * that means we have missed a write-write conflict, as it should have conflicted with this transaction.
+     */
+    private Optional<MismatchedValue> checkForWriteWriteConflicts(String table, WorkloadCell cell) {
+        ValueAndTimestamp previous = fetchValueFromView(table, cell, readView);
+        ValueAndTimestamp latest = fetchValueFromView(table, cell, latestView);
+
+        if (!previous.equals(latest)) {
```

Comment:
The comparison fails, which is the expected case, as we need to handle the situation when we do have no timestamp, such as when the value is null! 

So in this case, if I started with `null`, and then issued a delete later on, we'd want want to fail, even though the values match, because only one has a timestamp.